src/main/scala/scalafix/internal/sbt/ScalafixInterface.scala

```diff
@@ -116,47 +116,60 @@ object ScalafixInterface {
     private lazy val _value = scala.util.Try(thunk())
     override def apply(): T = _value.get
   }
+  private val fromToolClasspathMemo: scala.collection.mutable.HashMap[
+    (String, Seq[ModuleID], Seq[Repository]),
+    ScalafixInterface
+  ] = scala.collection.mutable.HashMap.empty
+
   def fromToolClasspath(
       scalafixBinaryScalaVersion: String,
       scalafixDependencies: Seq[ModuleID],
       scalafixCustomResolvers: Seq[Repository],
       logger: Logger = ConsoleLogger(System.out)
   ): () => ScalafixInterface =
     new LazyValue({ () =>
-      if (scalafixBinaryScalaVersion.startsWith("3"))
-        logger.error(
-          "To use Scalafix on Scala 3 projects, you must unset `scalafixBinaryScalaVersion`. " +
-            "Rules such as ExplicitResultTypes requiring the project version to match the Scalafix " +
-            "version are unsupported for the moment."
-        )
-      else if (scalafixBinaryScalaVersion == "2.11")
-        logger.error(
-          "Scala 2.11 is no longer supported. Please downgrade to the final version supporting " +
-            "it: sbt-scalafix 0.10.4."
-        )
-      val callback = new ScalafixLogger(logger)
-      val scalafixArguments = ScalafixAPI
-        .fetchAndClassloadInstance(
+      fromToolClasspathMemo.getOrElseUpdate(
```

Comment:
As this will be racy and the cost of loading stuff below is high (involving IO), I believe we should favor a Map implementation with blocking methods over one with only pure CAS methods. Unless I am mistaken, `ConcurrentHashMap.computeIfAbsent` does provide the guarantee that the block is never called more than once per key, while nothing in the Scala collections does. Could you update the implementation? Thanks!

Reply:
Updated to the existing `scalafix.internal.sbt.BlockingCache` which uses a `java.util.Collections.synchronizedMap` which should prevent race conditions.  Since the cardinality is low and the key access is only on scalafix on a per-project basis, this should also not block very often at all.

Reply:
Perfect, I had forgotten about that one!